### ISSUE TITLE
Extension: Pipe output to command palette in Terminal

### ISSUE THREAD

**Issue by zadjii-msft** [microsoft/terminal#8797]
Okay, this is gonna sound crazy so bear with me. I was switching git branches, and thought, "I really do not know how to use `less` well, and `git branch | $ grep foo` each time is annoying". 

Then I thought, we've got this great command palette for filtering a list of _things_. 

_As an extension_, what if we could pipe a list of items to the terminal, and then have the terminal open up the command palette with those items populated in it. Then selecting one of those items would _insert that text_?

So something like:

```
git branch | wt -w 0 select-list --prefix "git checkout "
```

would pipe to the current window, into the command palette. `--prefix "git checkout "` would indicate "use this text as a prefix for whatever the selected item was".

So hitting enter on one of these items, like "dev/migrie/foo" would `SendInput` `git checkout dev/migrie/foo` to the terminal.

This encapsulates a bunch of things:
* extensions need to be able to add their own subcommands to the commandline parser
  * they also need to be able to parse their own args
* extensions need to be able to open a ~command palette~ FilterableListView with whatever they want (a set of actions in this case. The actions would all be {sendInput(prefix+item), name=item})

**Comment by onein528** [comment]
I'd like to ask, does this repository also include the source code for Command Prompt ?

**Comment by DHowett** [comment]
No, and in an issue that’s not related to Command Prompt is not an appropriate place to ask :smile:

**Comment by zadjii-msft** [comment]
Hokay so it's not this simple. `wt` is a WINDOWS subsystem application, so even when it is piped to, we can't `GetStdHandle(STD_IN)`/`ReadFile` the input piped to it. 

We'd need the proposed `wtc` as a helper here. `wtc select-list` could check if there's input piped to it, and then use the Remoting lib to pass that to the monarch to pass to the active window...


WAIT what

`windowsterminal.exe`, when running unpackaged, accepts input from the commandline perfectly fine. But `wt` doesn't? Is this because there's the `wt.exe` shim in the middle that's not passing the input through to `windowsterminal.exe`? Oh it might be

**Comment by zadjii-msft** [comment]
hackathon week is my favorite week

![select-list-000](https://user-images.githubusercontent.com/18356694/127558472-7e0e5ae9-a20e-4a6a-8a6c-4199ba68d722.gif)

**Comment by zadjii-msft** [comment]
Remaining TODOS:
* [ ] generic todos throughout code - resources, etc. 
* [ ] after selecting/dismissing this menu, return to the normal list of commands in the palette. Right now it leaves them behind!
* [ ] Maybe _don't_ overload the `Command` ctor like I did?
* [ ] add a `-t,--trim` parameter, to remove whitespace on either side of the command
* [ ] add short form, `sl` (?)
* [ ] If `select-list` is the only action, and `-w` wasn't provided, then _default to `-w 0`_, and DONT ADD `new-tab`.  So `dir | wt sl` will by default run in the current terminal
* [ ] add a `--suffix` arg as well
* [ ] Ignore empty lines entirely.
* [ ] trim `\r`s off too. Don't want to be executing these things immediately!
  * [ ] maybe add a commandline arg to append the <kbd>enter</kbd> so the user can just execute immediately?
* [ ] `--icon` to set an icon for the commands? That seems wack though. If only we could font-fallback to nerd-fonts for their git symbol.
* [ ] Make sure that this can't actually get parsed by json. Or should it?
* [ ] Should `wt sl --option "foo" --option "bar"` etc work as expected? Like, should we actually allow that param on the commandline, and only use Stdin when it's actually there? (When would this be useful...?)
  * `wt sl -o cmd -o wsl -o pwsh --prefix "wt -w 0 nt"`? something like that?

<hr>

\<unbelievably wack idea>

`sendInput("dir /b /ad | wt sl --prefix cd")` bound to a key is fun. Maybe next year I'll think of a way to auto-bind that as a nested command, so you can navigate your directory tree with the command palette lol.

Maybe that would work as a combo with #9994 and #5970

```json
{
  "command": {
    "action": "executeActions",
    "subcommand": "cd",
    "actions": [
	  { "action":"sendInput", "input": "dir /b /ad | wt sl --prefix cd --suffix \"; wt cd\"\n"}
	]
  },
  "name": "change directory..."
},
```

so that would cause `wt cd` to input `dir /b /ad | wt sl --prefix cd --suffix "; wt cd"\n` to the terminal. We'd open up the command palette with the list of directories, and upon selecting one, we'd navigate to that directory, and _immediately do another `wt cd`_, bringing up the menu again. Hitting esc would dismiss this palette.

**Comment by cpriest** [comment]
Just a thought on this.  fzf already does this very well and works everywhere.

**Comment by zadjii-msft** [comment]
@cpriest Yea, I was mostly just messing around with this to see if this particular combination of things would work directly in the Terminal. It was more of an experiment than anything else. Almost more as a thought experiment more than anything else, but having the actual prototype in front of me was fun to play with and see what else we could do with something similar.

**Comment by zadjii-msft** [comment]
As a possibly related combination of wack ideas:

AS mentioned in #4719

> Another really useful plugin would be a Text Editor that could be used for profiles and scripts. The data would be piped to/from the Console for reading and saving

wack idea:

```type foo.txt | wt open-scratch```

where `open-scratch` will read from stdin. 

Though, it does seem easier to just 

```
wt edit foo.txt
```

---

This all might be _fully_ impossible with the new process model v3 as of 1.18. The wt that's run is a different process than the current one, so it'd be hard to plumb the HANDLE to the right place...
